🐛 bcrypt causes Edge Runtime compatibility issues in NextAuth middleware

**Describe the bug**
bcrypt dependency causes runtime errors when using NextAuth middleware in Next.js Edge Runtime due to native node-gyp bindings that are not supported in Edge Runtime environment.

**To Reproduce**
Steps to reproduce the behavior:

1. Run Next.js application with NextAuth middleware configured
2. Navigate to protected routes that trigger middleware
3. Observe Edge Runtime errors related to bcrypt native bindings
4. See error: "Module not found: Can't resolve 'bcrypt'" or similar node-gyp related errors

**Expected behavior**
Authentication should work seamlessly in Edge Runtime without dependency conflicts.

**Environment:**
- OS: Linux (WSL2)
- Browser: All browsers affected  
- Next.js: 15.4.4
- NextAuth: 5.0.0-beta.29
- bcrypt: 6.0.0

**Additional context**
The issue stems from bcrypt requiring native Node.js bindings (node-gyp compilation) which are incompatible with Next.js Edge Runtime. The middleware.ts exports NextAuth's auth function to Edge Runtime, causing the compatibility conflict.

This affects:
- `auth.ts:5` - bcrypt import for password verification
- `backend/auth/usecases/UserUsecase.ts:4` - bcrypt import for password hashing
- All test files importing bcrypt
- NextAuth middleware running in Edge Runtime

**Solution**
Replace bcrypt with bcryptjs (pure JavaScript implementation) for Edge Runtime compatibility.

--- a/app/api/auth/duplicates/route.ts
+++ b/app/api/auth/duplicates/route.ts
@@ -3,14 +3,19 @@
 import { NextRequest, NextResponse } from 'next/server';
 import { CheckUserIdDuplicateUseCase } from '@/backend/auth/usecases/UserUsecase';
 import { PrUserRepository } from '@/backend/common/infrastructures/repositories/PrUserRepository';
-import { CheckDuplicateRequestDto, CheckDuplicateResponseDto } from '@/backend/auth/dtos/UserDto';
+import {
+  CheckDuplicateRequestDto,
+  CheckDuplicateResponseDto,
+} from '@/backend/auth/dtos/UserDto';
 
 export async function POST(request: NextRequest) {
   try {
     const body: CheckDuplicateRequestDto = await request.json();
 
     const userRepository = new PrUserRepository();
-    const checkDuplicateUseCase = new CheckUserIdDuplicateUseCase(userRepository);
+    const checkDuplicateUseCase = new CheckUserIdDuplicateUseCase(
+      userRepository
+    );
 
     const result = await checkDuplicateUseCase.execute(body);
 

--- a/app/api/auth/duplicates/route.ts
+++ b/app/api/auth/duplicates/route.ts
@@ -1,0 +1,34 @@
+// ABOUTME: API endpoint to check if a user ID already exists in the system
+// ABOUTME: Used during registration flow to provide real-time duplicate validation
+import { NextRequest, NextResponse } from 'next/server';
+import { CheckUserIdDuplicateUseCase } from '@/backend/auth/usecases/UserUsecase';
+import { PrUserRepository } from '@/backend/common/infrastructures/repositories/PrUserRepository';
+import { CheckDuplicateRequestDto, CheckDuplicateResponseDto } from '@/backend/auth/dtos/UserDto';
+
+export async function POST(request: NextRequest) {
+  try {
+    const body: CheckDuplicateRequestDto = await request.json();
+
+    const userRepository = new PrUserRepository();
+    const checkDuplicateUseCase = new CheckUserIdDuplicateUseCase(userRepository);
+
+    const result = await checkDuplicateUseCase.execute(body);
+
+    if (!result.success) {
+      return NextResponse.json({ error: result.error }, { status: 400 });
+    }
+
+    const response: CheckDuplicateResponseDto = {
+      exists: result.exists!,
+      message: result.exists ? '이미 존재하는 아이디입니다.' : undefined,
+    };
+
+    return NextResponse.json(response, { status: 200 });
+  } catch (error) {
+    console.error('Duplicate check API error:', error);
+    return NextResponse.json(
+      { error: '서버 오류가 발생했습니다.' },
+      { status: 500 }
+    );
+  }
+}

--- a/auth.ts
+++ b/auth.ts
@@ -2,8 +2,6 @@
 // ABOUTME: Handles login using userId and password with bcrypt verification
 import NextAuth from 'next-auth';
 import Credentials from 'next-auth/providers/credentials';
-import bcrypt from 'bcrypt';
-import prisma from '@/libs/prisma';
 
 export const { handlers, signIn, signOut, auth } = NextAuth({
   providers: [
@@ -19,42 +17,6 @@ export const { handlers, signIn, signOut, auth } = NextAuth({
           label: '비밀번호',
           placeholder: '비밀번호',
         },
-      },
-      authorize: async (credentials) => {
-        if (!credentials?.id || !credentials?.password) {
-          return null;
-        }
-
-        try {
-          const user = await prisma.user.findUnique({
-            where: {
-              userId: credentials.id as string,
-            },
-          });
-
-          if (!user) {
-            return null;
-          }
-
-          const isValidPassword = await bcrypt.compare(
-            credentials.password as string,
-            user.password
-          );
-
-          if (!isValidPassword) {
-            return null;
-          }
-
-          return {
-            id: user.id.toString(),
-            name: user.name,
-            userId: user.userId,
-            role: user.role,
-          };
-        } catch (error) {
-          console.error('Auth error:', error);
-          return null;
-        }
       },
     }),
   ],

--- a/backend/auth/dtos/UserDto.ts
+++ b/backend/auth/dtos/UserDto.ts
@@ -20,3 +20,18 @@ export interface UserCreationResult {
   user?: CreateUserResponseDto;
   error?: string;
 }
+
+export interface CheckDuplicateRequestDto {
+  userId: string;
+}
+
+export interface CheckDuplicateResponseDto {
+  exists: boolean;
+  message?: string;
+}
+
+export interface DuplicateCheckResult {
+  success: boolean;
+  exists?: boolean;
+  error?: string;
+}

--- a/backend/auth/usecases/UserUsecase.ts
+++ b/backend/auth/usecases/UserUsecase.ts
@@ -1,9 +1,9 @@
 // ABOUTME: User use cases containing business logic for user operations
 // ABOUTME: Handles validation, password hashing, and user creation workflows
 
-import bcrypt from 'bcrypt';
+import bcrypt from 'bcryptjs';
 import { IUserRepository } from '../../common/domains/repositories/IUserRepository';
-import { CreateUserRequestDto, UserCreationResult } from '../dtos/UserDto';
+import { CreateUserRequestDto, UserCreationResult, CheckDuplicateRequestDto, DuplicateCheckResult } from '../dtos/UserDto';
 
 export class CreateUserUseCase {
   private userRepository: IUserRepository;
@@ -65,6 +65,39 @@ export class CreateUserUseCase {
       return {
         success: false,
         error: '사용자 생성 중 오류가 발생했습니다.',
+      };
+    }
+  }
+}
+
+export class CheckUserIdDuplicateUseCase {
+  private userRepository: IUserRepository;
+
+  constructor(userRepository: IUserRepository) {
+    this.userRepository = userRepository;
+  }
+
+  async execute(request: CheckDuplicateRequestDto): Promise<DuplicateCheckResult> {
+    // Validate required field
+    if (!request.userId) {
+      return {
+        success: false,
+        error: '아이디를 입력해주세요.',
+      };
+    }
+
+    try {
+      const existingUser = await this.userRepository.findByUserId(request.userId);
+      
+      return {
+        success: true,
+        exists: !!existingUser,
+      };
+    } catch (error) {
+      console.error('Duplicate check error:', error);
+      return {
+        success: false,
+        error: '중복 확인 중 오류가 발생했습니다.',
       };
     }
   }

--- a/backend/auth/usecases/UserUsecase.ts
+++ b/backend/auth/usecases/UserUsecase.ts
@@ -3,7 +3,12 @@
 
 import bcrypt from 'bcryptjs';
 import { IUserRepository } from '../../common/domains/repositories/IUserRepository';
-import { CreateUserRequestDto, UserCreationResult, CheckDuplicateRequestDto, DuplicateCheckResult } from '../dtos/UserDto';
+import {
+  CreateUserRequestDto,
+  UserCreationResult,
+  CheckDuplicateRequestDto,
+  DuplicateCheckResult,
+} from '../dtos/UserDto';
 
 export class CreateUserUseCase {
   private userRepository: IUserRepository;
@@ -77,7 +82,9 @@ export class CheckUserIdDuplicateUseCase {
     this.userRepository = userRepository;
   }
 
-  async execute(request: CheckDuplicateRequestDto): Promise<DuplicateCheckResult> {
+  async execute(
+    request: CheckDuplicateRequestDto
+  ): Promise<DuplicateCheckResult> {
     // Validate required field
     if (!request.userId) {
       return {
@@ -87,8 +94,10 @@ export class CheckUserIdDuplicateUseCase {
     }
 
     try {
-      const existingUser = await this.userRepository.findByUserId(request.userId);
-      
+      const existingUser = await this.userRepository.findByUserId(
+        request.userId
+      );
+
       return {
         success: true,
         exists: !!existingUser,

--- a/package.json
+++ b/package.json
@@ -15,8 +15,9 @@
     "@prisma/client": "^6.13.0",
     "@prisma/extension-accelerate": "^2.0.2",
     "@tanstack/react-query": "^5.84.1",
+    "@types/bcryptjs": "^2.4.6",
     "axios": "^1.11.0",
-    "bcrypt": "^6.0.0",
+    "bcryptjs": "^3.0.2",
     "next": "15.4.4",
     "next-auth": "^5.0.0-beta.29",
     "react": "19.1.0",
@@ -28,7 +29,6 @@
     "@eslint/eslintrc": "^3",
     "@testing-library/jest-dom": "^6.6.4",
     "@testing-library/react": "^16.3.0",
-    "@types/bcrypt": "^6.0.0",
     "@types/jest": "^30.0.0",
     "@types/node": "^20",
     "@types/react": "^19",
@@ -43,6 +43,7 @@
     "jest-environment-jsdom": "^30.0.5",
     "prettier": "^3.6.2",
     "prisma": "^6.13.0",
+    "tsx": "^4.20.3",
     "typescript": "^5",
     "typescript-eslint": "^8.38.0"
   }

--- a/tests/e2e/auth-flow.test.ts
+++ b/tests/e2e/auth-flow.test.ts
@@ -1,7 +1,7 @@
 // ABOUTME: End-to-end tests for complete authentication flow
 // ABOUTME: Tests full user registration and login process with NextAuth integration
 import { signIn, signOut } from 'next-auth/react';
-import bcrypt from 'bcrypt';
+import bcrypt from 'bcryptjs';
 
 const mockSignIn = signIn as jest.MockedFunction<typeof signIn>;
 const mockSignOut = signOut as jest.MockedFunction<typeof signOut>;

--- a/tests/helpers/test-db.ts
+++ b/tests/helpers/test-db.ts
@@ -1,6 +1,6 @@
 // ABOUTME: Database test utilities for creating and cleaning test data
 // ABOUTME: Provides helper functions for test user creation and database cleanup
-import bcrypt from 'bcrypt';
+import bcrypt from 'bcryptjs';
 import prisma from '@/libs/prisma';
 
 // Type assertion for mocked Prisma to avoid type conflicts in tests

--- a/tests/integration/auth-api.test.ts
+++ b/tests/integration/auth-api.test.ts
@@ -1,7 +1,7 @@
 // ABOUTME: Integration tests for authentication API endpoints
 // ABOUTME: Tests user registration and login flows with clean architecture layers
-import { CreateUserUseCase } from '@/backend/auth/usecases/UserUsecase';
-import { CreateUserRequestDto } from '@/backend/auth/dtos/UserDto';
+import { CreateUserUseCase, CheckUserIdDuplicateUseCase } from '@/backend/auth/usecases/UserUsecase';
+import { CreateUserRequestDto, CheckDuplicateRequestDto } from '@/backend/auth/dtos/UserDto';
 
 // Mock the entire PrUserRepository module
 jest.mock(
@@ -19,6 +19,7 @@ jest.mock(
 
 describe('Auth API Integration Tests', () => {
   let createUserUseCase: CreateUserUseCase;
+  let checkDuplicateUseCase: CheckUserIdDuplicateUseCase;
   let mockUserRepository: jest.Mocked<{
     findByUserId: jest.Mock;
     findById: jest.Mock;
@@ -32,6 +33,7 @@ describe('Auth API Integration Tests', () => {
     } = require('@/backend/common/infrastructures/repositories/PrUserRepository');
     mockUserRepository = new PrUserRepository();
     createUserUseCase = new CreateUserUseCase(mockUserRepository);
+    checkDuplicateUseCase = new CheckUserIdDuplicateUseCase(mockUserRepository);
   });
 
   describe('CreateUserUseCase', () => {
@@ -139,6 +141,88 @@ describe('Auth API Integration Tests', () => {
         expect(typeof message).toBe('string');
         expect(message.length).toBeGreaterThan(0);
       });
+    });
+  });
+
+  describe('CheckUserIdDuplicateUseCase', () => {
+    it('should check if user exists', async () => {
+      const existingUser = {
+        id: 'user123',
+        userId: 'existinguser',
+        password: 'hashedpassword',
+        name: '기존유저',
+        role: 'student',
+        point: 0,
+        streak: 0,
+        createdAt: new Date(),
+      };
+
+      mockUserRepository.findByUserId.mockResolvedValue(existingUser);
+
+      const request: CheckDuplicateRequestDto = { userId: 'existinguser' };
+      const result = await checkDuplicateUseCase.execute(request);
+
+      expect(result.success).toBe(true);
+      expect(result.exists).toBe(true);
+      expect(mockUserRepository.findByUserId).toHaveBeenCalledWith('existinguser');
+    });
+
+    it('should check if user does not exist', async () => {
+      mockUserRepository.findByUserId.mockResolvedValue(null);
+
+      const request: CheckDuplicateRequestDto = { userId: 'newuser' };
+      const result = await checkDuplicateUseCase.execute(request);
+
+      expect(result.success).toBe(true);
+      expect(result.exists).toBe(false);
+      expect(mockUserRepository.findByUserId).toHaveBeenCalledWith('newuser');
+    });
+
+    it('should validate required fields for duplicate check', async () => {
+      const invalidRequest: CheckDuplicateRequestDto = { userId: '' };
+      const result = await checkDuplicateUseCase.execute(invalidRequest);
+
+      expect(result.success).toBe(false);
+      expect(result.error).toBe('아이디를 입력해주세요.');
+      expect(mockUserRepository.findByUserId).not.toHaveBeenCalled();
+    });
+
+    it('should handle database errors during duplicate check', async () => {
+      mockUserRepository.findByUserId.mockRejectedValue(new Error('Database connection error'));
+
+      const request: CheckDuplicateRequestDto = { userId: 'testuser' };
+      const result = await checkDuplicateUseCase.execute(request);
+
+      expect(result.success).toBe(false);
+      expect(result.error).toBe('중복 확인 중 오류가 발생했습니다.');
+    });
+  });
+
+  describe('Duplicate Check API Response Formats', () => {
+    it('should format duplicate check success response correctly', () => {
+      const availableResponse = {
+        exists: false,
+      };
+
+      const duplicateResponse = {
+        exists: true,
+        message: '이미 존재하는 아이디입니다.',
+      };
+
+      expect(availableResponse.exists).toBe(false);
+      expect(availableResponse).not.toHaveProperty('message');
+
+      expect(duplicateResponse.exists).toBe(true);
+      expect(duplicateResponse.message).toBe('이미 존재하는 아이디입니다.');
+    });
+
+    it('should format duplicate check error response correctly', () => {
+      const errorResponse = {
+        error: '아이디를 입력해주세요.',
+      };
+
+      expect(errorResponse.error).toBe('아이디를 입력해주세요.');
+      expect(typeof errorResponse.error).toBe('string');
     });
   });
 });

--- a/tests/integration/auth-api.test.ts
+++ b/tests/integration/auth-api.test.ts
@@ -1,7 +1,13 @@
 // ABOUTME: Integration tests for authentication API endpoints
 // ABOUTME: Tests user registration and login flows with clean architecture layers
-import { CreateUserUseCase, CheckUserIdDuplicateUseCase } from '@/backend/auth/usecases/UserUsecase';
-import { CreateUserRequestDto, CheckDuplicateRequestDto } from '@/backend/auth/dtos/UserDto';
+import {
+  CreateUserUseCase,
+  CheckUserIdDuplicateUseCase,
+} from '@/backend/auth/usecases/UserUsecase';
+import {
+  CreateUserRequestDto,
+  CheckDuplicateRequestDto,
+} from '@/backend/auth/dtos/UserDto';
 
 // Mock the entire PrUserRepository module
 jest.mock(
@@ -164,7 +170,9 @@ describe('Auth API Integration Tests', () => {
 
       expect(result.success).toBe(true);
       expect(result.exists).toBe(true);
-      expect(mockUserRepository.findByUserId).toHaveBeenCalledWith('existinguser');
+      expect(mockUserRepository.findByUserId).toHaveBeenCalledWith(
+        'existinguser'
+      );
     });
 
     it('should check if user does not exist', async () => {
@@ -188,7 +196,9 @@ describe('Auth API Integration Tests', () => {
     });
 
     it('should handle database errors during duplicate check', async () => {
-      mockUserRepository.findByUserId.mockRejectedValue(new Error('Database connection error'));
+      mockUserRepository.findByUserId.mockRejectedValue(
+        new Error('Database connection error')
+      );
 
       const request: CheckDuplicateRequestDto = { userId: 'testuser' };
       const result = await checkDuplicateUseCase.execute(request);

--- a/tests/unit/auth.test.ts
+++ b/tests/unit/auth.test.ts
@@ -1,6 +1,9 @@
 // ABOUTME: Unit tests for authentication utilities and bcrypt functions
 // ABOUTME: Tests password hashing, validation, and user data processing
-import bcrypt from 'bcrypt';
+import bcrypt from 'bcryptjs';
+import { CheckUserIdDuplicateUseCase } from '@/backend/auth/usecases/UserUsecase';
+import { IUserRepository } from '@/backend/common/domains/repositories/IUserRepository';
+import { User } from '@/backend/common/domains/entities/User';
 
 describe('Authentication Utils', () => {
   describe('Password Hashing', () => {
@@ -72,6 +75,68 @@ describe('Authentication Utils', () => {
     it('should accept valid passwords', () => {
       const validPassword = 'password123';
       expect(validPassword.length).toBeGreaterThanOrEqual(6);
+    });
+  });
+
+  describe('CheckUserIdDuplicateUseCase', () => {
+    let mockRepository: jest.Mocked<IUserRepository>;
+    let useCase: CheckUserIdDuplicateUseCase;
+
+    beforeEach(() => {
+      mockRepository = {
+        findByUserId: jest.fn(),
+        findById: jest.fn(),
+        create: jest.fn(),
+      };
+      useCase = new CheckUserIdDuplicateUseCase(mockRepository);
+    });
+
+    it('should return validation error when userId is empty', async () => {
+      const result = await useCase.execute({ userId: '' });
+
+      expect(result.success).toBe(false);
+      expect(result.error).toBe('아이디를 입력해주세요.');
+      expect(mockRepository.findByUserId).not.toHaveBeenCalled();
+    });
+
+    it('should return exists: true when user exists', async () => {
+      const mockUser: User = {
+        id: '1',
+        userId: 'existinguser',
+        password: 'hashedpassword',
+        name: '기존유저',
+        role: 'student',
+        point: 0,
+        streak: 0,
+        createdAt: new Date(),
+      };
+      mockRepository.findByUserId.mockResolvedValue(mockUser);
+
+      const result = await useCase.execute({ userId: 'existinguser' });
+
+      expect(result.success).toBe(true);
+      expect(result.exists).toBe(true);
+      expect(mockRepository.findByUserId).toHaveBeenCalledWith('existinguser');
+    });
+
+    it('should return exists: false when user does not exist', async () => {
+      mockRepository.findByUserId.mockResolvedValue(null);
+
+      const result = await useCase.execute({ userId: 'newuser' });
+
+      expect(result.success).toBe(true);
+      expect(result.exists).toBe(false);
+      expect(mockRepository.findByUserId).toHaveBeenCalledWith('newuser');
+    });
+
+    it('should handle repository errors gracefully', async () => {
+      mockRepository.findByUserId.mockRejectedValue(new Error('Database error'));
+
+      const result = await useCase.execute({ userId: 'testuser' });
+
+      expect(result.success).toBe(false);
+      expect(result.error).toBe('중복 확인 중 오류가 발생했습니다.');
+      expect(mockRepository.findByUserId).toHaveBeenCalledWith('testuser');
     });
   });
 });

--- a/tests/unit/auth.test.ts
+++ b/tests/unit/auth.test.ts
@@ -130,7 +130,9 @@ describe('Authentication Utils', () => {
     });
 
     it('should handle repository errors gracefully', async () => {
-      mockRepository.findByUserId.mockRejectedValue(new Error('Database error'));
+      mockRepository.findByUserId.mockRejectedValue(
+        new Error('Database error')
+      );
 
       const result = await useCase.execute({ userId: 'testuser' });
 

